### ISSUE TITLE
variable translation to object block (or something along the lines)

### ISSUE THREAD

**Issue by segersniels** [hashicorp/terraform#17603]
Now let me say, finding a title for the following issue (or feature request) wasn't exactly as easy as you'd think. Let's jump right into the feature as I have no clue to explain myself.

Let's say for example I am rocking the following (dumbed down) configuration for an `aws_cloudfront_distribution` module:

```bash
resource "aws_cloudfront_distribution" "distribution" {
  origin {
    domain_name = "some-bucket"
    origin_id   = "bucket-Origin"
    origin_path = "/"
  }
  enabled = true
  comment = "${var.description}"
  default_root_object = "index.html
  default_cache_behavior {
    allowed_methods = ["DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT"]
    cached_methods = ["GET", "HEAD"]
    target_origin_id = "${var.name}-Origin"
    ...
  }
  ...
}
```

When applying the above you'd get an `origin` linked to the standard behaviour `*`, redirecting all the variations in paths (/foo, /bar, etc.) to the origin. But the idea is to map certain *"behaviours"* to different `origins` through the use of a module. So being able to configure the use of the module per project. 

What I'd like to accomplish is something like the following:

```bash
variable "extra_origins" { type = "list" }
variable "extra_cache_behaviors" { type = "list" }

resource "aws_cloudfront_distribution" "distribution" {
  origin {
    domain_name = "some-bucket"
    origin_id   = "bucket-Origin"
    origin_path = "/"
  }  

  ${var.extra_origins}
  ${var.extra_cache_behaviours}
  
  default_cache_behavior {
    allowed_methods  = ["DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT"]
    cached_methods   = ["GET", "HEAD"]
    target_origin_id = "${var.name}-Origin"
    ...
  }
}
```

```bash
extra_origins = [
  {
    origin {
      domain_name = "api.foo.com"
      origin_id   = "foo-Origin"
      origin_path = "/v1/foo"
    }
  },
  {
    origin {
      domain_name = "api.foo.com"
      origin_id   = "bar-Origin"
      origin_path = "/v2/bar"
    }
  }
]

extra_cache_behaviours = [
  {
    cache_behavior {
      allowed_methods  = ["DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT"]
      cached_methods   = ["GET", "HEAD"]
      target_origin_id = "foo-Origin"
      ...
      path_pattern = "/foo"
    }
  },
  {
    cache_behavior {
      allowed_methods  = ["DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT"]
      cached_methods   = ["GET", "HEAD"]
      target_origin_id = "bar-Origin"
      ...
      path_pattern = "/bar"
    }
  }
]
```

In other words, I'm either trying to do something that shouldn't be done or I found something that could be of use for other use cases as well as it would open up a lot of customization. I might just be overthinking this as there might already be an alternative way to accomplish this. Every input is welcome!

**TL;DR**: I have no idea if this makes any sense, but I'd like Terraform to translate a list into multiple `origin` and `cache_behavior` blocks for an `aws_cloudfront_distribution`. To be able to reuse and individualy reconfigure a module for different projects that require different `redirect` behaviours.

**Comment by apparentlymart** [comment]
Hi @segersniels! Thanks for this feature request.

I think you are describing here the same use-case as #7034, so I'm going to close this just to consolidate the discussion over there. We are hoping to include a feature for this in the next major release, and there are some notes about what it might look like in the other issue.
